src/test/java/com/reservation/payment/domain/PaymentTest.java

```diff
@@ -0,0 +1,172 @@
+
+        PaymentFailedAt failedAt2 = new PaymentFailedAt(LocalDateTime.now().plusSeconds(1));
+        FailureReason reason2 = new FailureReason("again");
+
+        payment.markAsFailed(failedAt1, reason1);
+
+        assertThrows(PaymentAlreadyFailedException.class,
+                () -> payment.markAsFailed(failedAt2, reason2));
+    }
+
+    @Test
+    void shouldNotAllowFailingAfterCompletion() {
+        Payment payment = new PaymentTestBuilder().build();
+        PaymentCompletedAt completedAt = new PaymentCompletedAt(LocalDateTime.now());
+        PaymentFailedAt failedAt = new PaymentFailedAt(LocalDateTime.now().plusSeconds(1));
+        FailureReason reason = new FailureReason("too late");
+
+        payment.markAsCompleted(completedAt);
+
+        assertThrows(PaymentAlreadyCompletedException.class,
+                () -> payment.markAsFailed(failedAt, reason));
+    }
+
+    @Test
+    void shouldNotAllowCompletingAfterFailure() {
+        Payment payment = new PaymentTestBuilder().build();
+        PaymentFailedAt failedAt = new PaymentFailedAt(LocalDateTime.now());
+        FailureReason reason = new FailureReason("too early");
+
+        PaymentCompletedAt completedAt = new PaymentCompletedAt(LocalDateTime.now().plusSeconds(1));
+
+        payment.markAsFailed(failedAt, reason);
+
+        assertThrows(PaymentAlreadyFailedException.class,
+                () -> payment.markAsCompleted(completedAt));
+    }
+
+
+    @Test
+    void shouldRegisterCompletedEvent() {
+        Payment payment = new PaymentTestBuilder().build();
+        PaymentCompletedAt completedAt = new PaymentCompletedAt(LocalDateTime.now());
+
+        payment.markAsCompleted(completedAt);
+
+        var events = payment.pullEvents();
+        assertEquals(1, events.size());
+        assertInstanceOf(PaymentCompletedEvent.class, events.getFirst());
+
+        PaymentCompletedEvent event = (PaymentCompletedEvent) events.getFirst();
+        assertEquals(payment.getPaymentId(), event.paymentId());
+        assertEquals(payment.getReservationId(), event.reservationId());
+        assertEquals(completedAt, event.completedAt());
+    }
+
+    @Test
+    void shouldRegisterFailedEvent() {
+        Payment payment = new PaymentTestBuilder().build();
+        PaymentFailedAt failedAt = new PaymentFailedAt(LocalDateTime.now());
+        FailureReason reason = new FailureReason("Too late");
+
+        payment.markAsFailed(failedAt, reason);
+
+        var events = payment.pullEvents();
+        assertEquals(1, events.size());
+        assertTrue(events.get(0) instanceof PaymentFailedEvent);
+
+        PaymentFailedEvent event = (PaymentFailedEvent) events.get(0);
+        assertEquals(reason, event.reason());
+        assertEquals(failedAt, event.failedAt());
+    }
+
+    @Test
+    void shouldReturnEmptyEventListIfNoEventsRegistered() {
+        Payment payment = new PaymentTestBuilder().build();
+        var events = payment.pullEvents();
```

Comment:
Tutaj też można zamienić `var`
```suggestion
        List<PaymentEvent> events = payment.pullEvents();
```